brian2/codegen/cpp_prefs.py

```diff
@@ -236,3 +239,87 @@ def update_for_cross_compilation(library_dirs, extra_compile_args,
         library_dirs += ['/lib32', '/usr/lib32']
         extra_compile_args += ['-m32']
         extra_link_args += ['-m32']
+
+
+_msvc_env = None
+def get_msvc_env():
+    global _msvc_env
+    from setuptools import msvc
+    import distutils
+    vcvars_loc = prefs['codegen.cpp.msvc_vars_location']
+    if vcvars_loc:
+        return vcvars_loc, None
+    arch_name = prefs['codegen.cpp.msvc_architecture']
+    if arch_name == '':
+        bits = struct.calcsize('P') * 8
+        if bits == 64:
+            arch_name = 'x86_amd64'
+        else:
+            arch_name = 'x86'
+    if _msvc_env is None:
+        try:
+            # This will fail on Python 2 with an AttributError
+            _msvc_env = msvc.msvc14_get_vc_env(arch_name)
+        except (distutils.errors.DistutilsPlatformError, AttributeError):
+            # Use the old way of searching for MSVC
+            # FIXME: Remove this when we remove Python 2 support
+            #        and require Visual Studio 2015?
+            for version in range(16, 8, -1):
+                fname = msvc.msvc9_find_vcvarsall(version)
+                if fname:
+                    vcvars_loc = fname
+                    break
+            if vcvars_loc == '':
+                raise IOError("Cannot find Microsoft Visual Studio, You "
+                              "can try to set the path to vcvarsall.bat "
+                              "via the codegen.cpp.msvc_vars_location "
+                              "preference explicitly.")
+    if vcvars_loc:
+        vcvars_cmd = '"{vcvars_loc}" {arch_name}'.format(vcvars_loc=vcvars_loc,
+                                                         arch_name=arch_name)
+    else:
+        vcvars_cmd = None
+    return _msvc_env, vcvars_cmd
+
+
+_compiler_supports_c99 = None
+def compiler_supports_c99():
+    global _compiler_supports_c99
+    if _compiler_supports_c99 is None:
+        if platform.system() == 'Windows':
+            fd, tmp_file = tempfile.mkstemp(suffix='.cpp')
+            os.write(fd, '''
+            #if _MSC_VER < 1800
+            #error
+            #endif
+            '''.encode())
+            os.close(fd)
+            msvc_env, vcvars_cmd = get_msvc_env()
+            if vcvars_cmd:
+                cmd = '{} && cl /E mytest.cpp > NUL 2>&1'.format(vcvars_cmd)
+            else:
+                os.environ.update(msvc_env)
+                cmd = 'cl /E mytest.cpp > NUL 2>&1'
+            return_value = os.system(cmd)
+            _compiler_supports_c99 = return_value == 0
+            os.remove(tmp_file)
+        else:
+            cmd = ('echo "#if (__STDC_VERSION__ < 199901L)\n#error\n#endif" | '
+                  'cc -xc -c - > /dev/null 2>&1')
+            return_value = os.system(cmd)
+            _compiler_supports_c99 = return_value == 0
+    return _compiler_supports_c99
```

Comment:
I guess this check doesn't take that long, but it always makes me a little uncomfortable having these sorts of things running every time.

Reply:
It is cached in the global `_compiler_supports_c99` variable.

Reply:
But not between runs?

Reply:
Ah, no, indeed. But then I don't know how to cache it between invocations of Python without making things complicated to detect changes in the compiler setup. The part that potentially takes some non-negligible time is finding/executing `vcvarsall.bat`, but we'd have to do this every time to notice a compiler update I guess?

Reply:
Executing ``vcvarsall.bat`` is probably slow, but maybe we can just use its exact path as a key? If that path hasn't changed you don't need to do the rest to check the C99 compatibility?

Reply:
I think finding `vcvarsall.bat` in the first place is slow... Also, the "modern version" of doing it (which is needed for the newer Visual Studio versions) doesn't actually return the path for the file but a dictionary with all the environment variables.